test/quantization/ptq/wrappers/test_quant_module_base.py

```diff
@@ -0,0 +1,209 @@
+# limitations under the License.
+
+"""
+Unit-tests for the abstract helper class **QuantModuleBase**.
+
+Because the class is abstract, the tests create a tiny concrete subclass
+(`DummyQM`) that:
+
+1. owns exactly **one** observer (`obs`)
+2. in `forward()` multiplies the input by 2.0 and passes it through `_fq`
+   so we can verify collection / fake-quant behaviour.
+
+The suite checks:
+
+* default mode is **NO_QUANT**
+* `enable_calibration()` resets the observer and switches mode
+* `_fq()` really collects in CALIB and fake-quantises in QUANT
+* `freeze_qparams()` disables the observer and populates cached q-params
+* `_make_obs()` merges overrides from a `QuantConfig`
+"""
+
+import math, torch, unittest
+from typing import Dict
+
+from tico.experimental.quantization.ptq.dtypes import DType
+from tico.experimental.quantization.ptq.mode import Mode
+from tico.experimental.quantization.ptq.observers import MinMaxObserver
+from tico.experimental.quantization.ptq.observers.affine_base import AffineObserverBase
+from tico.experimental.quantization.ptq.observers.base import ObserverBase
+from tico.experimental.quantization.ptq.observers.ema import EMAObserver
+from tico.experimental.quantization.ptq.qscheme import QScheme
+from tico.experimental.quantization.ptq.quant_config import QuantConfig
+from tico.experimental.quantization.ptq.wrappers.quant_module_base import (
+    QuantModuleBase,
+)
+
+
+# concrete toy subclass
+class DummyQM(QuantModuleBase):
+    def __init__(self, qcfg: QuantConfig | None = None):
+        super().__init__(qcfg)
+        self.obs = self._make_obs("act")
+
+    def forward(self, x):
+        return self._fq(x * 2.0, self.obs)
+
+    def _all_observers(self):
+        return (self.obs,)
+
+
+class TestQuantModuleBase(unittest.TestCase):
+    def setUp(self):
+        torch.manual_seed(0)
+        self.x = torch.randn(16, 4)
+        self.qm = DummyQM()  # default uint8
+
+    def test_mode_cycle(self):
+        self.assertIs(self.qm._mode, Mode.NO_QUANT)
+
+        self.qm.enable_calibration()
+        self.assertIs(self.qm._mode, Mode.CALIB)
+        # observer reset to +inf / -inf
+        assert isinstance(self.qm.obs, AffineObserverBase)
+        self.assertTrue(math.isinf(self.qm.obs.min_val.item()))
+
+        self.qm.freeze_qparams()
+        self.assertIs(self.qm._mode, Mode.QUANT)
+        self.assertTrue(self.qm.obs.has_qparams)
+
+    def test_fq_collect_and_quantize(self):
+        # CALIB pass – observer should collect
+        self.qm.enable_calibration()
+        _ = self.qm(self.x)
+        assert isinstance(self.qm.obs, AffineObserverBase)
+        lo = self.qm.obs.min_val.item()
+        hi = self.qm.obs.max_val.item()
+        self.assertLess(lo, hi)  # stats updated
+
+        # QUANT pass – output must differ from FP32
+        self.qm.freeze_qparams()
+        q_out = self.qm(self.x)
+        fp_out = self.x * 2.0
+        self.assertFalse(torch.allclose(q_out, fp_out))
```

Comment:
```suggestion
        self.assertFalse(torch.equal(q_out, fp_out))
```
(optional)
How about to use `torch.equal` here?
Is there no possibility for the case when the results are in the tolerance range, since quantization is done **well**?

Reply:
> How about to use torch.equal here?

`torch.equal` always results in `False` on float values. So, `allclose` is appropriate in this case.

> Is there no possibility for the case when the results are in the tolerance range, since quantization is done well?

I think it won't be. Let's fix it when it is the case.